protocol/message_sender.go

```diff
@@ -0,0 +1,250 @@
+package protocol
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+	"github.com/libp2p/go-msgio"
+
+	pb "github.com/libp2p/go-smart-record/protocol/pb"
+)
+
+// ErrReadTimeout is an error that occurs when no message is read within the timeout period.
+var ErrReadTimeout = fmt.Errorf("timed out reading response")
+
+// messageSenderImpl is responsible for sending requests and messages to peers
+type messageSenderImpl struct {
+	host      host.Host // the network services we need
+	smlk      sync.Mutex
+	strmap    map[peer.ID]*peerMessageSender
+	protocols []protocol.ID
+}
+
+func (m *messageSenderImpl) streamDisconnect(ctx context.Context, p peer.ID) {
+	m.smlk.Lock()
+	defer m.smlk.Unlock()
+	ms, ok := m.strmap[p]
+	if !ok {
+		return
+	}
+	delete(m.strmap, p)
+
+	// Do this asynchronously as ms.lk can block for a while.
+	go func() {
+		if err := ms.lk.Lock(ctx); err != nil {
+			return
+		}
+		defer ms.lk.Unlock()
+		ms.invalidate()
+	}()
+}
+
+// SendRequest sends out a request
+func (m *messageSenderImpl) SendRequest(ctx context.Context, p peer.ID, pmes *pb.Message) (*pb.Message, error) {
+
+	ms, err := m.messageSenderForPeer(ctx, p)
+	if err != nil {
+		log.Debugw("request failed to open message sender", "error", err, "to", p)
+		return nil, err
+	}
+
+	rpmes, err := ms.SendRequest(ctx, pmes)
+	if err != nil {
+		return nil, err
+	}
+
+	return rpmes, nil
+}
+
+// SendMessage sends out a message
+func (m *messageSenderImpl) SendMessage(ctx context.Context, p peer.ID, pmes *pb.Message) error {
+
+	ms, err := m.messageSenderForPeer(ctx, p)
+	if err != nil {
+		log.Debugw("message failed to open message sender", "error", err, "to", p)
+		return err
+	}
+
+	if err := ms.SendMessage(ctx, pmes); err != nil {
+		log.Debugw("message failed", "error", err, "to", p)
+		return err
+	}
+
+	return nil
+}
+
+func (m *messageSenderImpl) messageSenderForPeer(ctx context.Context, p peer.ID) (*peerMessageSender, error) {
+	m.smlk.Lock()
+	ms, ok := m.strmap[p]
+	if ok {
+		m.smlk.Unlock()
+		return ms, nil
+	}
+	ms = &peerMessageSender{p: p, m: m, lk: newCtxMutex()}
+	m.strmap[p] = ms
+	m.smlk.Unlock()
+
+	if err := ms.prepOrInvalidate(ctx); err != nil {
+		m.smlk.Lock()
+		defer m.smlk.Unlock()
+
+		if msCur, ok := m.strmap[p]; ok {
```

Comment:
You are right! That is why `streamDisconnect` is not used anywhere (not even exposed in the package). I used it for testing purposes and forgot to remove. Removing it now to avoid confusion. 

As we use a request/response pattern, the server handling the request is responsible for closing the stream gracefully after sending a response at `handleNewStream`. Otherwise, streams are closed with errors if something fails.